src/stream/src/executor/backfill.rs

```diff
@@ -111,28 +127,33 @@ where
 
     #[try_stream(ok = Message, error = StreamExecutorError)]
     async fn execute_inner(mut self) {
-        // The primary key columns, in the output columns of the table scan.
-        let pk_in_output_indices = self.table.pk_in_output_indices().unwrap();
-        let pk_order = self.table.pk_serializer().get_order_types();
+        // The primary key columns, in the output columns of the upstream_table scan.
+        let pk_in_output_indices = self.upstream_table.pk_in_output_indices().unwrap();
+
+        let pk_order = self.upstream_table.pk_serializer().get_order_types();
 
-        let table_id = self.table.table_id().table_id;
+        let dist_key_in_pk = self.dist_key_in_pk;
+
+        let upstream_table_id = self.upstream_table.table_id().table_id;
 
         let mut upstream = self.upstream.execute();
 
         // Poll the upstream to get the first barrier.
         let first_barrier = expect_first_barrier(&mut upstream).await?;
         let init_epoch = first_barrier.epoch.prev;
+        self.state_table.init_epoch(first_barrier.epoch);
 
         // If the barrier is a conf change of creating this mview, we follow the procedure of
         // backfill. Otherwise, it means we've recovered and we can forward the upstream messages
         // directly.
         let to_create_mv = first_barrier.is_newly_added(self.actor_id);
```

Comment:
I got another reason why we need to persist multiple rows into the state for the backfill executor. Suppose a mv-on-mv backfill is finished, but we scale the upstream mv. When we check the state to determine whether or not to proceed with backfilling, it is possible that we would find an empty state for a backfill executor (the new one after scale), if we only persist one row for each backfill executor.

Reply:
> I got another reason why we need to persist multiple rows into the state for the backfill executor. Suppose a mv-on-mv backfill is finished, but we scale the upstream mv. When we check the state to determine whether or not to proceed with backfilling, it is possible that we would find an empty state for a backfill executor (the new one after scale), if we only persist one row for each backfill executor.

Oh I guess empty state indicates we are not done with backfill. Fair point. We should persist row position for all vnodes.

But regardless we still can't scale the backfill right even *after backfilling is finished* because of https://github.com/risingwavelabs/risingwave/pull/9752#discussion_r1196527071 below.